### PR TITLE
Ignore timeouts while busy

### PR DESCRIPTION
**Pull request by EmielBruijntjes**
Design choice:

- The DNS-CPP library first _reads_ data from the UDP socket
- And then it sets up an idle-handler to process this data in a later iteration of the event loop
- Reason: OS-level UDP sockets have a limited buffer size, and we want to read the data as soon as possible, without doing the (potential slow) parsing + call back to userspace

Question:

- What if a DNS lookup times out after 60 seconds (this is configurable)
- While we do not yet have processed the answer
- Even though that answer is in the internal buffer of DNS-CPP?

Current situation:

- The timeout is passed to userspace as a timeout or SERVFAIL error
- Which is not entirely correct, because we did receive a valid answer
- And the problem was not because of a SERVFAIL
- But was caused by the CPU that could not process all the responses

~Partial solution (this project):~

- ~When the request times out~
- ~We first check if there are no internal buffers left~
- ~If we do have internal buffers, we postpone the timeout until all buffers are empty~
 
~Potential new problem:~

- ~If the buffer never gets empty (because DNS requests are started all the time)~
- ~We will never see DNS timeouts, because by the time of the timeout,~
- ~The buffer already holds other requests~

Better solution:

- When the request times out
- We look at the unprocessed messages in the buffer
- More specifically: we check when those messages were received
- If there are any messages in the buffers that were received _before_ the expire-time of the lookup
- We now wait a little longer before we expire the lookup
- Until we only have messages in the buffers that were received after the expire time

### CODE REVIEW

File: src/udp.cpp [hunk 1]
```diff
@@ -171,13 +184,38 @@ void Udp::notify()
...
+    // we are going to check if the udp socket is readable
+    pollfd info;
+    
+    // check for readable
+    info.fd = _fd;
+    info.events = POLLIN;
+    info.revents = 0;     
+
+    // poll the fd with no timeout
+    return poll(&info, 1, 0) > 0;
```
Comment (rwols) [inline src/udp.cpp:135]: Wouldn't this have to go through the Loop class?

**Comment by EmielBruijntjes** [comment]
Note that because of the last item ('potential new problem') on the list, I am not sure if we should be merging this.

File: src/udp.cpp (hunk 1 above)
Comment (EmielBruijntjes) [reply src/udp.cpp:135]: No, it is just a trick to check if there is something in the _internal_ buffer, without actually reading it.

File: src/udp.cpp [hunk 2]
```diff
@@ -259,6 +224,11 @@ bool Udp::send(const Ip &ip, const Query &query)
  */
 bool Udp::send(const struct sockaddr *address, size_t size, const Query &query)
 {
+    // before we send we empty the buffer (this is an obscure optimization to deal
+    // with use-cases where many domains are resolved, without any calls to the event
+    // loop in between, the socket could then already hold some responses)
+//    if (readable()) notify();
```
Comment (RafalGoslawski) [inline src/udp.cpp]: Why is this line commented out?

**Review by RafalGoslawski — APPROVED** [review]
looks good to me

**Pull request merged**